feat: Refactor Generator

FrankEssenberger opened the pull request:
## Context

This PR cleans up the generation of the `VdmServiceMetadata` so all the entities, function-imports and complex types. The main changes are:
- The edmx data is just read and stored to `root`. The  needed parts (compex type,entity...) are parsed later in the methods where the entity,complex type and function imports are generated. With this the intermediate edmx step is hidden from the generator stronger and only one split between v2 and v4 is needed.
- The `service-generator` contains the only switch for v2/v4
- The parsing of edmx and generation of vdm has been split (`parser` and `service-vdm` folders)
- I changed the  service parser to a class because it needed state like the formatter, options and so, which is more natural then passing so many arguments.
- In the generation implementation there are always three things for v2,v4,common: `entity-vdm.ts`, `function-import-vdm.ts` and `complex-type-vdm.ts` which take care of generating these VDM properties. The old `edmx-to-vdm` was simply to large files. 
- I also created multiple util files holding some scattered methods to make the actual classes not so big.
- I put type definitions in a `type` file so that we can use `import type` in case of cyclic dependencies easily

In order to guarantee backwards compatibility I moved  the old stuff as it is was in a `legacy` folder and deprecated everything. From the new generator I only exported the `service-generator` to root level which could be useful do some programmatic generation without writing stuff to the disc like the generator does. However, all the internal stuff stays hidden so that we are free to change things there.

## Definition of Done

Please consider all items and remove only if not applicable.

- [x] Tests created/adjusted for your changes.
- [x] Release notes updated.
  * Provide sufficient context so that each entry can be understood on its own.
  * Be specific about names of functions, classes, modules, etc.
  * Describe when or where this is relevant
  * Use indicative and present tense. For example, write "Provide function `name` that does X in order to Y" over "Now X can be done by calling a new function".
- [x] PR title adheres to [conventional commit guidelines](https://www.conventionalcommits.org) (please note that only `fix:` and `feat:` will end up in the release notes)
- [x] If applicable: Properly documented (JSDoc of public API)
- [x] If applicable: Check if `node run doc` still works.

FrankEssenberger commented:
@jjtang1985 @marikaner So I stop now with the moving around of things. Please have a look.

marikaner commented on CHANGELOG.md:
```diff
@@ -22,7 +22,7 @@
 ## Improvements
 
--
+- Refactored the service vdm generation and deprecated old flow.
```
Minor: Use imperative and maybe make it less general. I guess we will need compatibility notes, but feel free to do this at the end.

FrankEssenberger replied:
I adjusted the changelog but it is really hard for me to write aktive and this imperative is almost impossible. I spend now 15 min on a few sentences. I am simply used to either a collective `we did ABC` or `ABC has been done` in phrasing some activities.

FrankEssenberger commented on packages/generator/src/parser/common/complex-type-parser.ts:
```diff
@@ -0,0 +1,98 @@
+
+const logger = createLogger({
+  package: 'generator',
+  messageContext: 'edmx-complex-type-parser'
```
I tried to align the file names

FrankEssenberger commented on packages/generator/src/parser/common/complex-type-parser.ts:
```diff
@@ -0,0 +1,98 @@
+  return forceArray(root.ComplexType).map(c => {
+    c.Property = forceArray(c.Property);
+    return c;
+  });
```
yes.

FrankEssenberger commented on packages/generator/src/parser/common/complex-type-parser.ts:
```diff
@@ -0,0 +1,98 @@
+export function transformComplexTypesBase(
+  complexTypes: EdmxComplexTypeBase[],
+  formatter: ServiceNameFormatter
+  // reservedNames: Set<string>
```
yes

FrankEssenberger commented on packages/generator/src/parser/common/complex-type-parser.ts:
```diff
@@ -0,0 +1,98 @@
+  const formattedTypes = complexTypes.reduce(
+    (formatted, c) => ({
+      ...formatted,
+      [c.Name]: formatter.originalToComplexTypeName(c.Name)
```
This was just copy paste from the old code [see here](https://github.com/SAP/cloud-sdk/blob/9d0256a71c4e4deaf96be8b87e8512d87c28b01c/packages/generator/src/parser/common/edmx-to-vdm.ts#L345-L347)

FrankEssenberger replied:
so I hope it is correct, but I do not really know :-).

FrankEssenberger commented on packages/generator/src/parser/common/edmx-types.ts:
```diff
@@ -0,0 +1,66 @@
+  PropertyRef: EdmxNamed[];
+}
+
+export interface EdmxEntityTypeBase<NavigationType> extends EdmxNamed {
```
good point. I apply the "Type" to "T" for all generic parameter.

FrankEssenberger commented on packages/generator/src/parser/common/entity-parser.ts:
```diff
@@ -0,0 +1,253 @@
+// TODO more elegant way to pass the type in?
+export function parseEntityTypesBase<NavigationType>(
+  root,
+  foo: NavigationType
```
sure the name of the parameter I already adjusted. In java there is the SomeThing.class which can be used for such things. For TS I do not know

FrankEssenberger commented on packages/generator/src/parser/common/entity-parser.ts:
```diff
@@ -0,0 +1,253 @@
+export function parseEntityTypesBase<NavigationType>(
+  root,
+  foo: NavigationType
+): EdmxEntityTypeBase<NavigationType>[] {
```
So on the when you parse the base type you do a ` e.NavigationProperty = forceArray(e.NavigationProperty);`. Since `root` has type any this works. However I wanted to indicate in the interface that the property is there and will be parsed. The type however differes between v2 and v4. Hence this generic form.

FrankEssenberger commented:
> I can't say, that I looked through everything as there are a lot of changes ;)
> But here are my 2 cents: I definitely like the changes, I think the split was unnatural and from what I can tell this is easier and more natural now. I left you some comments for details, but I also have some general remarks:
> 
> 1. We should be careful with the usage of VDM. From a product perspective VDM denotes the pregenerated OData clients for SAP S/4HANA. This generator is therefore not about VDM, but about OData clients. With this spirit, I know in the beginning we prefixed many of the types with `Vdm` therefore for this I would say, what is done is done, but we should not introduce more usage of that. That said, I have two sub points:
>    a. think `service-vdm` is not a good name nor place for what it contains. Shouldn't this be below `parser`?
>    b. Many of the files now have the suffix, `-vdm`. I would suggest to just drop it.
> 2. `ServiceVdmGenerator` - I don't really see a major benefit to having a class instead of the function that was there before (maybe renamed). If I saw it correctly it is always used in combination with `getAllServices` which indicates that the created instance of the class itself is not really necessary.

To 1: I removed the vdm and made a folder `edmx-to-vdm` containing now the parts going from edmx to the objects called `VdmTypes`. Since the edmx-to-vdm was also there in the past in single files I think this is fair. 

To2: I added the class mainly for the state of options and the name formatter which you would have to pass around the methods, or use the global scope of a function to wrap it  which we do in other places, but in such a case I find it more transparent to use a class to hold this state.

FrankEssenberger commented on packages/generator/src/service-vdm/vdm-types.ts:
```diff
@@ -80,16 +85,24 @@ export interface VdmComplexType {
+  returnTypeEdmx: string;
+}
+
+export interface VdmFunctionImport extends VdmFunctionWithoutReturnType {
```
Ok I used Omit to represent the two cases.

FrankEssenberger commented on packages/generator/src/service-vdm/vdm-types.ts:
```diff
@@ -80,16 +85,24 @@ export interface VdmComplexType {
   description: string;
 }
 
+export interface VdmFunctionImportReturnTypeNotParsed
```
see below

FrankEssenberger commented:
I did generate the test-serives and git showed me no difference. At least the refactoring has not changed something seriously.

FrankEssenberger commented on packages/generator/src/service-vdm/service-vdm-generator.ts:
```diff
@@ -0,0 +1,204 @@
+}
+
+/**
+ * @deprecated Sinve version 1.25.0. Use the ServiceParser class instead
```
thanks.

FrankEssenberger commented on packages/generator/src/service-vdm/v4/complex-type-vdm.ts:
```diff
@@ -0,0 +1,27 @@
+  };
+}
+
+export function getComplexTypesV4(
```
The parsing of the complex types in the edmx is different. For v4  the joinTypesWithBaseTypes is executed. I could of course hand over the edmx type to the transform method and remove the two implementations.

FrankEssenberger replied:
but then it would so be so symmetric in the generation step: 
```
  const functionImportsWithoutReturnType = isV2Metadata(serviceMetadata.edmx)
      ? generateFunctionImportsV2(serviceMetadata, formatter)
      : generateFunctionImportsV4(serviceMetadata, formatter);

    const complexTypes = isV2Metadata(serviceMetadata.edmx)
      ? generateComplexTypesV2(serviceMetadata, formatter)
      : generateComplexTypesV4(serviceMetadata, formatter);

    const entities = isV2Metadata(serviceMetadata.edmx)
      ? generateEntitiesV2(serviceMetadata, complexTypes, formatter)
      : generateEntitiesV4(serviceMetadata, complexTypes, formatter);
```

marikaner commented on packages/generator/src/service-generator.ts:
```diff
@@ -0,0 +1,196 @@
+    serviceDefinitionPaths: ServiceDefinitionPaths
+  ): ServiceMetadata {
+    const serviceMetadata: ServiceMetadata = {
+      // TODO: pass parameter
```
TODO?

FrankEssenberger replied:
hm, to old to bad written. Forgott already what I wanted to do. Note to myself. Either make a better TODO or do it right away.

marikaner commented on packages/generator/src/service-name-formatter.ts:
```diff
@@ -24,13 +38,26 @@ export class ServiceNameFormatter {
     complexTypeNames: string[],
     functionImportNames: string[]
+  );
+  constructor(
+    entitySetNames?: string[],
+    complexTypeNames?: string[],
+    functionImportNames?: string[]
```
```suggestion
  constructor(
    entitySetNames: string[] = [],
    complexTypeNames: string[] = [],
    functionImportNames: string[] = []
```

marikaner commented on packages/generator/src/service-name-formatter.ts:
```diff
@@ -24,13 +38,26 @@ export class ServiceNameFormatter {
   ) {
     // Here we assume that entitysets and complextypes cannot have the same original name
-    [...entitySetNames, ...complexTypeNames].forEach(
+    [...(entitySetNames || []), ...(complexTypeNames || [])].forEach(
```
Togher with the suggestion above:
```suggestion
    [...entitySetNames, ...complexTypeNames].forEach(
```

FrankEssenberger commented on packages/generator/src/edmx-parser/common/edmx-parser.ts:
```diff
@@ -0,0 +1,34 @@
+  }));
+}
+
+// TODO more elegant way to pass the type in?
+export function parseEntityTypesBase<NavigationT>(
+  root,
+  navigationType: NavigationT
+): EdmxEntityTypeBase<NavigationT>[] {
```
I see you point if we would have a navigation base type one would extend, then there could be some benefit. But lets make it any and cast on the return,.

FrankEssenberger commented on packages/generator/src/edmx-parser/legacy/common/edmx-parser.ts:
```diff
@@ -3,19 +3,23 @@
 const logger = createLogger({
   package: 'generator',
-  messageContext: 'edmx-parser'
+  messageContext: 'edmx-edmx-parser'
```
no I guess a global replace all error.

FrankEssenberger commented on packages/generator/src/edmx-parser/edmx-file-reader.ts:
```diff
@@ -0,0 +1,88 @@
+  messageContext: 'edmx-edmx-parser'
+});
+
+export interface EdmxMetadataBase {
```
yes good catch.

FrankEssenberger commented on packages/generator/src/generator.ts:
```diff
@@ -331,8 +331,8 @@ function projectOptions(): ProjectOptions {
 
-function readServices(options: GeneratorOptions): VdmServiceMetadata[] {
-  const services = parseAllServices(options);
+function generateServices(options: GeneratorOptions): VdmServiceMetadata[] {
```
parse it is.

FrankEssenberger commented on packages/generator/src/edmx-parser/index.ts:
```diff
@@ -0,0 +1,2 @@
+/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
+export * from './legacy';
```
ok

marikaner commented on packages/generator/src/edmx-parser/v2/edmx-parser.ts:
```diff
@@ -0,0 +1,42 @@
+}
+
+export function parseFunctionImports(root): EdmxFunctionImport[] {
+  return forceArray(root.EntityContainer.FunctionImport).map(f => {
+    f.Parameter = forceArray(f.Parameter);
+    return f;
+  });
```
I guess this was overlooked.

marikaner commented on packages/generator/src/edmx-parser/v4/edmx-parser.ts:
```diff
@@ -0,0 +1,77 @@
+    filterEnumProperties(entityTypes, parseEnumTypes(root)),
+    joinEntityTypes
+  );
+} // TODO: Filters enum properties as long as those are not supported
+function filterEnumProperties(
```
Same here

marikaner commented on packages/generator/src/edmx-parser/v4/edmx-parser.ts:
```diff
@@ -0,0 +1,77 @@
+}
+
+export function parseFunctions(root): EdmxFunction[] {
+  return forceArray(root.Function).map(f => {
+    f.Parameter = forceArray(f.Parameter);
+    return f;
+  });
```
And here

marikaner commented on packages/generator/src/edmx-to-vdm/description-util.ts:
```diff
@@ -0,0 +1,87 @@
+  property: EdmxProperty,
+  swaggerProperty?: SwaggerProperty
+): string {
+  let desc = '';
+  if (property['sap:quickinfo']) {
+    desc = property['sap:quickinfo'];
+  } else if (property['sap:label']) {
+    desc = property['sap:label'];
+  } else if (swaggerProperty && swaggerProperty.title) {
+    desc = swaggerProperty.title;
+  }
```
don't forget me

marikaner commented on packages/generator/src/edmx-to-vdm/description-util.ts:
```diff
@@ -0,0 +1,87 @@
+  entity: JoinedEntityMetadata<EdmxEntitySetBase, any>,
+  className: string
+): string {
+  if (entity.entityType['sap:label']) {
+    return entity.entityType['sap:label'];
+  }
+  return entity.swaggerDefinition && entity.swaggerDefinition.title
+    ? entity.swaggerDefinition.title
+    : toTitleFormat(className);
```
or me

marikaner commented on packages/generator/src/service-generator.ts:
```diff
@@ -0,0 +1,196 @@
+    const formatter = new ServiceNameFormatter();
+
+    /*
+    Do function imports before complex types so that function with name "createSomething" gets a nicer name, because the builder function of a complex type would also be called `create${ComplexTypeName}`.
```
?

FrankEssenberger replied:
No you are absolutely right. The approach of skipping just the factory name is much cleaner. I will change this. This also gets rid of the duplicate return type in the funciton import.

marikaner commented on packages/generator/src/service-generator.ts:
```diff
@@ -0,0 +1,196 @@
+    /*
+    Do function imports before complex types so that function with name "createSomething" gets a nicer name, because the builder function of a complex type would also be called `create${ComplexTypeName}`.
+    */
+    const functionImportsWithoutReturnType = isV2Metadata(serviceMetadata.edmx)
+      ? generateFunctionImportsV2(serviceMetadata, formatter)
+      : generateFunctionImportsV4(serviceMetadata, formatter);
+
+    const complexTypes = isV2Metadata(serviceMetadata.edmx)
+      ? generateComplexTypesV2(serviceMetadata, formatter)
+      : generateComplexTypesV4(serviceMetadata, formatter);
+
+    const entities = isV2Metadata(serviceMetadata.edmx)
+      ? generateEntitiesV2(serviceMetadata, complexTypes, formatter)
+      : generateEntitiesV4(serviceMetadata, complexTypes, formatter);
```
?

FrankEssenberger replied:
I have overseen this. Created two methods in the edmx-to-vdm of the v2/v4 case

marikaner commented on packages/generator/src/service-generator.ts:
```diff
@@ -0,0 +1,143 @@
+import { apiBusinessHubMetadata } from './edmx-parser/swagger/swagger-util';
+import { VdmServiceMetadata, VdmServicePackageMetaData } from './vdm-types';
+import { isV2Metadata } from './edmx-to-vdm/edmx-to-vdm-util';
+import { getServiceEntitiesV2 } from './edmx-to-vdm/v2/service-entities';
+import { getServiceEntitiesV4 } from './edmx-to-vdm/v4/service-entities';
```
```suggestion
import { getServiceEntitiesV2 } from './edmx-to-vdm/v2';
import { getServiceEntitiesV4 } from './edmx-to-vdm/v4';
```

marikaner commented on packages/generator/src/vdm-types.ts:
```diff
@@ -83,11 +88,12 @@ export interface VdmComplexType {
   parametersTypeName: string;
   functionName: string;
   description: string;
+  returnTypeEdmx: string;
```
With this we have duplicate information - VdmFunctionImportReturnTypeCategory.

FrankEssenberger replied:
also a symptom of the bad return type handling. Now removed after the factory name for the comple type is left.

FrankEssenberger commented on packages/generator/src/service-generator.ts:
```diff
@@ -0,0 +1,196 @@
+import { parseReturnTypes } from './edmx-to-vdm/common/function-import';
+import { generateFunctionImportsV2 } from './edmx-to-vdm/v2/function-import';
+import { generateEntitiesV2 } from './edmx-to-vdm/v2/entity';
+import { generateComplexTypesV2 } from './edmx-to-vdm/v2/complex-type';
```
added the index files

FrankEssenberger commented on packages/generator/src/edmx-to-vdm/common/function-import.ts:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    const isCollectionReturnType = isCollection(returnType);
+    const edmxType = findEdmType(returnType);
```
True. I refacctor it also in the other files like edm-types.ts

FrankEssenberger commented on packages/generator/src/edmx-parser/swagger/swagger-util.ts:
```diff
@@ -0,0 +1,51 @@
+/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
```
ok

FrankEssenberger commented on packages/generator/src/edmx-to-vdm/common/function-import.ts:
```diff
@@ -0,0 +1,182 @@
+  functionImports: Omit<VdmFunctionImport, 'returnType'>[],
+  entities: VdmEntity[],
+  complexTypes: VdmComplexType[]
+): VdmFunctionImport[] {
```
parses now only the return type.

FrankEssenberger commented on packages/generator/src/service-vdm/service-vdm-generator.ts:
```diff
@@ -0,0 +1,204 @@
+
+/**
+ * @deprecated Sinve version 1.25.0. Use the ServiceParser class instead
+ * @param options Generator options *
```
removed

FrankEssenberger commented on packages/generator/src/service-vdm/service-vdm-generator.ts:
```diff
@@ -0,0 +1,204 @@
+}
+
+/**
+ * @deprecated Sinve version 1.25.0. Use the ServiceParser class instead
```
Method is exposed now and not the class. Deprecated removed.

FrankEssenberger commented:
Hi Marika, sorry I did not address all your comments directly but the github ui somehow collapsed most of them.... 
Now I hope I have addressed everything.

FrankEssenberger commented:
I also checked and there seems to be no changes in the generated test entities induces by the refactoring.

marikaner commented on packages/generator/src/edmx-to-vdm/common/function-import.ts:
```diff
@@ -0,0 +1,175 @@
+  if (returnType.startsWith('Edm.')) {
+    return returnType;
+  }
+  return undefined;
```
```suggestion
```

FrankEssenberger commented:
And 95 comments later it is merged. Uh, uh.
